site/content/posts/2023-08-09-v0.25.0-impersonation-proxy-with-external-certs.md

```diff
@@ -0,0 +1,337 @@
+
+```shell
+# From https://projectcontour.io/getting-started/
+$ kubectl apply \
+    --filename https://projectcontour.io/quickstart/contour.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+# Verify that the Contour pods are ready
+$ kubectl get pods \
+    --namespace projectcontour \
+    --output wide \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Install Pinniped’s local-user-authenticator and add some sample users (see https://pinniped.dev/docs/tutorials/concierge-only-demo/ for more details).
+
+```shell
+# Install Pinniped's local-user-authenticator
+$ kubectl apply \
+    --filename https://get.pinniped.dev/v0.25.0/install-local-user-authenticator.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+# Create a local user "pinny" with password "password123" and group "group-for-mtls".
+# Each secret in this namespace acts like a user definition.
+$ kubectl create secret generic pinny \
+    --namespace local-user-authenticator \
+    --from-literal=groups=group-for-mtls \
+    --from-literal=passwordHash=$(htpasswd -nbBC 10 x password123 | sed -e "s/^x://") \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+# We'll need the CA bundle of the local-user-authenticator service to configure the Concierge's WebhookAuthenticator.
+# Just make sure this next command does print out the TLS secret, which can take a few seconds to generate.
+$ kubectl get secret local-user-authenticator-tls-serving-certificate \
+    --namespace local-user-authenticator \
+    --output jsonpath={.data.caCertificate} \
+    --kubeconfig kind-with-contour.kubeconfig.yaml \
+    | tee local-user-authenticator-ca.pem.b64
+```
+
+Install Pinniped’s Concierge:
+
+```shell
+$ kubectl apply \
+    --filename https://get.pinniped.dev/v0.25.0/install-pinniped-concierge-crds.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+
+$ kubectl apply \
+    --filename https://get.pinniped.dev/v0.25.0/install-pinniped-concierge-resources.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Install `cert-manager`:
+
+```shell
+$ kubectl apply \
+    --filename https://github.com/cert-manager/cert-manager/releases/download/v1.12.3/cert-manager.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Configure a `cert-manager` certificate for the impersonation proxy to serve TLS.
+
+Note that this section bootstraps a CA `Issuer` used to issue leaf certificates that can be used to serve TLS.
+For more information on this, see the [cert-manager docs](https://cert-manager.io/docs/configuration/selfsigned/#bootstrapping-ca-issuers).
+The `Certificate` with name `impersonation-serving-cert` will generate the leaf certificate used by the impersonation proxy to serve TLS.
+
+```shell
+$ cat << EOF > self-signed-cert.yaml
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: cert-manager
+---
+apiVersion: cert-manager.io/v1
+kind: ClusterIssuer
+metadata:
```

Comment:
So we have 

- `ClusterIssuer` named `selfsigned-cluster-issuer`
- `ClusterIssuer` named `my-ca-issuer`

Can we describe what each one is/does?  I can't clearly state one vs the other.

If I took a stab at it:

 The `ClusterIssuer` named `my-ca-issuer` will be used to create several `Certificate` resources.  First, we will create the `Certificate` called `my-selfsigned-ca` (and will store the actual certificate data in the `Secret` called `self-signed-ca-for-kind-testing`).  

We will later retrieve the `Secret` called `self-signed-ca-for-kind-testing` so that we can add the CA to the Pinniped Concierge's `CredentialIssuer` resource so it can be advertised and used to verify TLS serving certificates.

Then, we will create the `ClusterIssuer` called `my-ca-issuer`.  We will reference the `Certificate` called `my-selfsigned-ca` via it's `Secret` named `self-signed-ca-for-kind-testing`.  This will allow us to use the CA to sign TLS serving certificates.

Tyeh, we will use the `ClusterIssuer` called `my-ca-issuer` to generate a `Certificate` that will be a TLS serving certificate called `impersonation-serving-cert`.  As before, the actual cerficiate will be stored in a kubernetes `Secret` which we will name `impersonation-proxy-tls-serving-cert`.

Finally, we will update the Pinniped Concierge's `CredentiaIissuer` resource to use the TLS serving certificate stored in the `Secret` called `impersonation-proxy-tls-serving-cert`.

If all goes well, the Impersonation Proxy endpoints will be served with a TLS serving certificate that can be validated by the CA certificate that generated it.

Reply:
Resolving locally....